### PR TITLE
Automatically decompress a compressed file with `FileService`

### LINKED ISSUE #3804: Dynamically decompress compressed files with `FileService`
Currently, `FileService` can serve compressed files with:
https://github.com/line/armeria/blob/5b384fbe27e7e6f9225d6db91cbb684d09dfbb5e/core/src/main/java/com/linecorp/armeria/server/file/FileServiceBuilder.java#L113
However, it serves the compressed file itself and does not decompress the compressed file if a client does not accept a compressed file.

If a user packages both compressed and decompressed files in a JAR file, it would have an impact on the overall JAR size.
It would be nice to add an option that is able to decompress compressed files.
```java
public FileServiceBuilder serveCompressedFiles(boolean serveCompressedFiles, boolean autoDecompression) {
    ...
}
```

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

A `FileService` serve compressed files depending on the accept-encoding
when `FileServiceBuilder.serveCompressedFiles(boolean)` is enabled.
If a compressed file such as `foo.json.gz` only exists, a request for
`foo.json` would be failed.

The size of uncompressed files is relatively big than compressed one.
Some users might not want to package the uncompressed files to a JAR file.

Modifications:

- Add `autoDecompress` option to `FileServiceBuilder`.
- Decompress a compressed file if a client wants to accept uncompressed content.
- Add `StaticHttpDecodedResponse` that decodes `HttpObject`s using
  a predefined `StreamDecoder`.
- Fix `FileService` to cache uncompressed files.
- Move some package-private classes to the internal packages.

Result:

- You can now decompress compressed files when using `FileService`
- Fixes #3804

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/server/file/FileService.java [hunk 1]
```diff
@@ -344,19 +372,40 @@ private HttpFile findFile(ServiceRequestContext ctx, HttpRequest req) {
...
 
         assert cache != null;
 
         final Executor executor = ctx.blockingTaskExecutor();
         final ByteBufAllocator alloc = ctx.alloc();
 
         return HttpFile.from(uncachedFile.aggregateWithPooledObjects(executor, alloc).thenApply(aggregated -> {
+            if (decompress && encoding != null) {
+                aggregated = decompress(aggregated, encoding, alloc);
+                if (aggregated.attributes().length() > config.maxCacheEntrySizeBytes()) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/file/FileService.java:402]: `aggregated.attributes()` is `null` when `aggregated` is `AggregatedHttpFile.nonExistent()`

**Review by minwoox — COMMENTED** [review]
Thanks! Left some nits. 😄

File: core/src/main/java/com/linecorp/armeria/internal/common/encoding/package-info.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
+/*
+ * Copyright 2016 LINE Corporation
```
[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/common/encoding/package-info.java:2]: 😱

File: core/src/main/java/com/linecorp/armeria/internal/common/encoding/StaticHttpDecodedResponse.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        this.contentType = contentType;
+    }
+
+    @Override
+    protected HttpObject filter(HttpObject obj) {
+        if (obj instanceof HttpData) {
+            return decoder.decode((HttpData) obj);
+        }
+
+        if (obj instanceof ResponseHeaders) {
```
[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/internal/common/encoding/StaticHttpDecodedResponse.java:55]: That might be better.

File: core/src/main/java/com/linecorp/armeria/server/file/FileService.java (hunk 1 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/server/file/FileService.java:402]: A non-existent file should not reach here and `decompress()` also checks `attributes`.
https://github.com/line/armeria/blob/c0acac56b8258f05b99f80aa8b11ad3959163ac3/core/src/main/java/com/linecorp/armeria/server/file/FileService.java#L428-L430
For clarification, let me add additional assertions.

**Review by trustin — COMMENTED** [review]
LGTM, but one question: Can't a user just compress the JAR file?

**Comment by ikhoon** [comment]
This PR aimed at reducing the bundled JavaScript file of the documentation service when packaging Armeria core module.
https://github.com/line/armeria/issues/3706#issuecomment-908998639

> Can't a user just compress the JAR file?

Did this mean to decompress a JAR file and serve in runtime?

**Comment by trustin** [comment]
> Did this mean to decompress a JAR file and serve in runtime?

No, the class loader will decompress it anyway when we try to read it.

But I agree that it makes sense that we ship pre-gzipped one in our JAR given most browsers accept gzipped resources.

**Review by trustin — APPROVED** [review]

**Review by minwoox — APPROVED** [review]
👍 👍 👍

**Pull request merged**